users/serializers.py

```diff
@@ -322,6 +328,92 @@ class Meta:
         )
 
 
+class ChangeEmailRequestCreateSerializer(serializers.ModelSerializer):
+    """Serializer for starting a user email change"""
+
+    user = serializers.HiddenField(default=serializers.CurrentUserDefault())
+    new_email = serializers.EmailField()
+    password = serializers.CharField(write_only=True)
+
+    def validate(self, attrs):
+        """Validate the change request"""
+        # verify no other user has this email address
+        errors = {}
+
+        user = attrs["user"]
+        new_email = attrs["new_email"]
+        password = attrs.pop("password")
+
+        if user.email == new_email:
+            # verify the user isn't trying to change their email to their current one
+            # this would indicate a programming error on the frontend if this request is allowed
+            errors["email"] = "Provided email address is same as your current one"
+        elif User.objects.filter(email=new_email).exists():
+            errors["email"] = "Invalid email address"
+
+        if errors:
+            raise serializers.ValidationError(errors)
+
+        # verify the password verifies for the current user
+        if not user.check_password(password):
+            raise serializers.ValidationError("Invalid Password")
+
+        return attrs
+
+    def create(self, validated_data):
+        """Create the email change request"""
+        change_request = super().create(validated_data)
+
+        verification_api.send_verify_email_change_email(
+            self.context["request"], change_request
+        )
+
+        return change_request
+
+    class Meta:
+        model = ChangeEmailRequest
+
+        fields = ("user", "new_email", "password")
+
+
+class ChangeEmailRequestUpdateSerializer(serializers.ModelSerializer):
+    """Serializer for confirming a user email change"""
+
+    confirmed = serializers.BooleanField()
+
+    @transaction.atomic
+    def update(self, instance, validated_data):
+        """Updates an email change request"""
+        if User.objects.filter(email=instance.new_email).exists():
+            log.debug(
+                "User %s tried to change email address to one already in use", instance
+            )
+            raise serializers.ValidationError("Unable to change email")
+
+        result = super().update(instance, validated_data)
+
+        # change request has been confirmed
+        if result.confirmed:
+            user = result.user
+            old_email = user.email
+            user.email = result.new_email
+            user.save()
+            # delete social_auth entry to avoid old email account access
+            try:
+                user_social_auth = UserSocialAuth.objects.get(uid=old_email, user=user)
+                user_social_auth.delete()
+            except UserSocialAuth.DoesNotExist:
+                pass
+            change_edx_user_email_async.delay(user.id)
```

Comment:
This needs some tests around it to confirm its doing the right thing.

Reply:
done.